python/ray/serve/application.py

```diff
@@ -0,0 +1,324 @@
+
+        if f:
+            yaml.safe_dump(json_str, f)
+        return yaml.safe_dump(json_str)
+
+    @classmethod
+    def from_yaml(cls, str_or_file: Union[str, TextIO]) -> "Application":
+        """Converts YAML data to deployments for an Application.
+
+        Takes in a string or a file pointer to a file containing deployment
+        definitions in YAML. These definitions are converted to a new
+        Application object containing the deployments.
+
+        To read from a file, use the following pattern:
+
+        with open("file_name.txt", "w") as f:
+            app = app.from_yaml(str_or_file)
+
+        Args:
+            str_or_file (Union[String, TextIO]): Either a string containing
+                YAML deployment definitions or a pointer to a file containing
+                YAML deployment definitions. The YAML format must adhere to the
+                ServeApplicationSchema JSON Schema defined in
+                ray.dashboard.modules.serve.schema. This function works with
+                Serve YAML config files.
+
+        Returns:
+            Application: a new Application object containing the deployments.
+        """
+
+        deployments_json = yaml.safe_load(str_or_file)
+        schema = ServeApplicationSchema.parse_obj(deployments_json)
+        return cls(schema_to_serve_application(schema))
+
+    def __getitem__(self, key: str):
+        """Fetch a deployment by name using dict syntax: app["name"]
+
+        Raises:
+            KeyError: if the name is not in this Application.
+        """
+
+        if key in self._deployments:
+            return self._deployments[key]
+        else:
+            raise KeyError(f'Serve application does not contain a "{key}" deployment.')
+
+    def __setitem__(self, key: str, value: Deployment):
+        """Set a deployment by name with dict syntax: app[name]=new_deployment
+
+        Use this to overwrite existing deployments.
+
+        Args:
+            key (String): name
+            value (Deployment): the deployment that the name maps to
+
+        Raises:
+            TypeError: if the key is not a String or the value is not a deployment.
+        """
+
+        if not isinstance(key, str):
+            raise TypeError(f"key should be a string, but got object of type {key}.")
+        elif not isinstance(value, Deployment):
+            raise TypeError(f"Got {type(Deployment)} for value. Expected deployment.")
+
+        self._deployments[key] = value
+
+    def __getattr__(self, name: str):
+        """Fetch a deployment by name using attribute syntax: app.name
+
+        Raises:
+            AttributeError: if the name is not in this Application.
+        """
+
+        if name in self._deployments:
+            return self._deployments[name]
+        else:
+            raise AttributeError(
+                f'Serve application does not contain a "{name}" deployment.'
+            )
+
```

Comment:
I think per @simon-mo's suggestion let's only do __getitem__ for now and not __getattr__ to avoid having two ways to do the same thing.